cpp/examples/parquet_inspect/parquet_inspect_utils.cpp

```diff
@@ -0,0 +1,420 @@
+  columns.emplace_back(make_index_column(num_row_groups, stream));
+
+  auto row_offsets_buffer =
+    rmm::device_buffer(row_group_row_offsets.data(), num_row_groups * sizeof(int64_t), stream);
+  auto row_counts_buffer =
+    rmm::device_buffer(row_group_row_counts.data(), num_row_groups * sizeof(int64_t), stream);
+  auto byte_offsets_buffer =
+    rmm::device_buffer(row_group_byte_offsets.data(), num_row_groups * sizeof(int64_t), stream);
+
+  columns.emplace_back(std::make_unique<cudf::column>(cudf::data_type{cudf::type_to_id<int64_t>()},
+                                                      num_row_groups,
+                                                      std::move(row_offsets_buffer),
+                                                      rmm::device_buffer{},
+                                                      0));
+  columns.emplace_back(std::make_unique<cudf::column>(cudf::data_type{cudf::type_to_id<int64_t>()},
+                                                      num_row_groups,
+                                                      std::move(row_counts_buffer),
+                                                      rmm::device_buffer{},
+                                                      0));
+  columns.emplace_back(std::make_unique<cudf::column>(cudf::data_type{cudf::type_to_id<int64_t>()},
+                                                      num_row_groups,
+                                                      std::move(byte_offsets_buffer),
+                                                      rmm::device_buffer{},
+                                                      0));
+
+  auto table = std::make_unique<cudf::table>(std::move(columns));
+
+  cudf::io::table_input_metadata out_metadata(table->view());
+  out_metadata.column_metadata[0].set_name("row group index");
+  out_metadata.column_metadata[1].set_name("row offsets");
+  out_metadata.column_metadata[2].set_name("row counts");
+  out_metadata.column_metadata[3].set_name("byte offsets");
+
+  auto const out_opts =
+    cudf::io::parquet_writer_options::builder(cudf::io::sink_info(output_filepath), table->view())
+      .metadata(out_metadata)
+      .build();
+  cudf::io::write_parquet(out_opts, stream);
+}
+
+void write_page_metadata(cudf::io::parquet::FileMetaData const& metadata,
+                         std::string const& output_filepath,
+                         rmm::cuda_stream_view stream)
+{
+  CUDF_FUNC_RANGE();
+
+  auto const num_columns    = metadata.row_groups.front().columns.size();
+  auto const num_row_groups = metadata.row_groups.size();
+
+  auto constexpr output_cols_per_column = 3;
+
+  std::vector<std::unique_ptr<cudf::column>> columns;
+  columns.emplace_back(make_index_column(num_row_groups, stream));
+
+  std::for_each(
+    thrust::counting_iterator<size_t>(0),
+    thrust::counting_iterator(num_columns),
+    [&](auto const col_idx) {
+      auto const [page_row_counts, page_row_offsets, page_byte_offsets, col_page_offsets] =
```

Comment:
can these go out of scope before we copy to device buffers? I don't see any syncs in `make_page_data_list_column`